src/logger/index.ts

```diff
@@ -56,28 +56,32 @@ function removeCircular(obj: any, refs: any[] = []): any {
   if (typeof obj !== "object" || !obj) {
     return obj;
   }
-  // If the object defines its own toJSON, prefer that.
-  if (obj.toJSON) {
+  // If the object defines its own toJSON method, use it.
+  if (obj.toJSON && typeof obj.toJSON === "function") {
     return obj.toJSON();
   }
+  // Only check for circularity among ancestors in the recursion stack.
   if (refs.includes(obj)) {
     return "[Circular]";
-  } else {
-    refs.push(obj);
   }
+  // Add the current object to the recursion stack.
+  refs.push(obj);
+
+  // If the object is an array, run circular check on each element.
   let returnObj: any;
   if (Array.isArray(obj)) {
-    returnObj = new Array(obj.length);
-  } else {
-    returnObj = {};
+    returnObj = obj.map((item) => removeCircular(item, refs));
   }
-  for (const k in obj) {
-    if (refs.includes(obj[k])) {
-      returnObj[k] = "[Circular]";
-    } else {
-      returnObj[k] = removeCircular(obj[k], refs);
+  // If the object is a Map, run circular check on each key and value.
```

Comment:
nitpick on the comment but we're not dealing with a js "Map" here, just an object